File: day_08.py
from pathlib import Path
from typing import List

import numpy as np

from constants import INPUTS_DIR, UTF_8

INPUT_PATH = Path(INPUTS_DIR) / "day-08.txt"
# INPUT_PATH = Path(INPUTS_DIR) / "example.txt"


def main1(grid: List[List[int]]) -> int:
    n_rows = len(grid)
    n_cols = len(grid[0])
    visible = np.zeros((n_rows, n_cols), dtype=bool)
    visible[0, :] = True
    visible[-1, :] = True
    visible[:, 0] = True
    visible[:, -1] = True
    for row in range(n_rows):
        # from left
        tallest = grid[row][0]
        for col in range(1, n_cols):
            if grid[row][col] > tallest:
                tallest = grid[row][col]
                visible[row, col] = True
        # from right
        tallest = grid[row][-1]
        for col in range(n_cols - 2, -1, -1):
            if grid[row][col] > tallest:
                tallest = grid[row][col]
                visible[row, col] = True
    for col in range(n_cols):
        # from up
        tallest = grid[0][col]
        for row in range(1, n_rows):
            if grid[row][col] > tallest:
                tallest = grid[row][col]
                visible[row, col] = True
        # from down
        tallest = grid[-1][col]
        for row in range(n_rows - 2, -1, -1):
            if grid[row][col] > tallest:
                tallest = grid[row][col]
                visible[row, col] = True
    return visible.sum()


def score(forest: List[List[int]], row: int, col: int) -> int:
    n_rows = len(forest)
    n_cols = len(forest[0])
    this_height = forest[row][col]
    # up
    for i in range(row - 1, -1, -1):
        if forest[i][col] >= this_height:
            up = row - i
            break
    else:
        up = row
    # down
    for i in range(row + 1, n_rows):
        if forest[i][col] >= this_height:
            down = i - row
            break
    else:
        down = n_rows - row - 1
    # left
    for i in range(col - 1, -1, -1):
        if forest[row][i] >= this_height:
            left = col - i
            break
    else:
        left = col
    # right
    for i in range(col + 1, n_cols):
        if forest[row][i] >= this_height:
            right = i - col
            break
    else:
        right = n_cols - col - 1
    return up * down * left * right


def main2(forest: List[List[int]]) -> int:
    highest = 0
    for i in range(1, len(forest) - 1):
        for j in range(1, len(forest[0]) - 1):
            highest = max(highest, score(forest, i, j))
    return highest


if __name__ == "__main__":
    with open(INPUT_PATH, "r", encoding=UTF_8) as f:
        forest_ = [list(map(int, line_.strip())) for line_ in f.readlines()]
    ans = main1(forest_)
    print("part 1:", ans)
    ans = main2(forest_)
    print("part 2:", ans)

File: kotlin/Day08.kt
fun solve1(forest: List<List<Int>>): Int {
    val n_rows = forest.size
    val n_cols = forest[0].size
    val visible = Array(n_rows){ Array(n_cols){ false } }
    for (i in 0 until n_rows) {
        visible[i][0] = true
        visible[i][n_cols - 1] = true
    }
    for (i in 0 until n_cols) {
        visible[0][i] = true
        visible[n_rows - 1][i] = true
    }
    // analyze from 4 sides
    for (row in 0 until n_rows) {
        // from left
        var tallest = forest[row][0]
        for (col in 1 until n_cols) {
            if (forest[row][col] > tallest) {
                tallest = forest[row][col]
                visible[row][col] = true
            }
        }
        // from right
        tallest = forest[row][n_cols - 1]
        for (col in (n_cols - 2) downTo 0) {
            if (forest[row][col] > tallest) {
                tallest = forest[row][col]
                visible[row][col] = true
            }
        }
    }
    for (col in 0 until n_cols) {
        // from up
        var tallest = forest[0][col]
        for (row in 1 until n_rows) {
            if (forest[row][col] > tallest) {
                tallest = forest[row][col]
                visible[row][col] = true
            }
        }
        // from down
        tallest = forest[n_rows - 1][col]
        for (row in (n_rows - 2) downTo 0) {
            if (forest[row][col] > tallest) {
                tallest = forest[row][col]
                visible[row][col] = true
            }
        }
    }
    return visible.sumOf{ it.count{ it } }
}


fun score(forest: List<List<Int>>, row: Int, col: Int): Int {
    val n_rows = forest.size
    val n_cols = forest[0].size
    val this_height = forest[row][col]
    // up
    var up_: Int? = null
    for (i in (row - 1) downTo 0) {
        if (forest[i][col] >= this_height) {
            up_ = row - i
            break
        }
    }
    val up = up_ ?: row
    // down
    var down_: Int? = null
    for (i in (row + 1) until n_rows) {
        if (forest[i][col] >= this_height) {
            down_ = i - row
            break
        }
    }
    val down = down_ ?: (n_rows - row - 1)
    // left
    var left_: Int? = null
    for (i in (col - 1) downTo 0) {
        if (forest[row][i] >= this_height) {
            left_ = col - i
            break
        }
    }
    val left = left_ ?: col
    // right
    var right_: Int? = null
    for (i in (col + 1) until n_cols) {
        if (forest[row][i] >= this_height) {
            right_ = i - col
            break
        }
    }
    val right = right_ ?: (n_cols - col - 1)
    return up * down * left * right
}


fun solve2(forest: List<List<Int>>): Int {
    var highest = 0
    for (i in 1 until (forest.size - 1)) {
        for (j in 1 until (forest[0].size - 1)) {
            val this_score = score(forest, i, j)
            if (this_score > highest) {
                highest = this_score
            }
        }
    }
    return highest
}


fun main() {
    val lines = mutableListOf<String>()
    while (true) {
        val line = readLine()
        if (line == null) {
            break
        }
        lines.add(line)
    }
    val forest = lines.map{ it.map{ it.digitToInt() } }
    val answer1 = solve1(forest)
    println("part 1: $answer1")
    val answer2 = solve2(forest)
    println("part 2: $answer2")
}
